Add support for more text_alignment values

vinc3m1 opened the pull request:
Fixes #516

Design notes:
- Did not use [Layout.Alignment](https://developer.android.com/reference/android/text/Layout.Alignment) because it doesn't differentiate between VIEW_START/END and TEXT_START/END
- Did not use `View.TEXT_ALIGNMENT_*`/[`android:textAlignment`](https://developer.android.com/reference/android/view/View#attr_android:textAlignment) because
  - Default value of `INHERIT` cannot be supported by Litho without adding this property to all Components which doesn't seem ideal
  - `GRAVITY` cannot be supported unless Litho adds a gravity prop to Text, which doesn't seem necessary
  - `LEFT` and `RIGHT` are not supported directly as values
- Proposed solution is to use our own Enum for `TextAlignment`.
- This deprecates the `textAlignment` prop for the new `alignment` prop that handles the new cases. Unfortunately the old prop's dependences on an Enum requires a migration.

vinc3m1 commented:
bump @astreet

colriot commented:
Hi @vinc3m1! Can you, please, share a snippet of the layout that can't be implemented in Litho?
Like, what you have in XML, basically. Just to make it more visual.

vinc3m1 commented:
According to https://developer.android.com/reference/android/view/View#attr_android:textAlignment:

Constant | Value | Description
-- | -- | --
center | 4 | Center the paragraph, for example: ALIGN_CENTER.
gravity | 1 | Default for the root view. The gravity determines the alignment, ALIGN_NORMAL, ALIGN_CENTER, or ALIGN_OPPOSITE, which are relative to each paragraph’s text direction.
inherit | 0 | Default.
textEnd | 3 | Align to the end of the paragraph, for example: ALIGN_OPPOSITE.
textStart | 2 | Align to the start of the paragraph, for example: ALIGN_NORMAL.
viewEnd | 6 | Align to the end of the view, which is ALIGN_RIGHT if the view’s resolved layoutDirection is LTR, and ALIGN_LEFT otherwise.
viewStart | 5 | Align to the start of the view, which is ALIGN_LEFT if the view’s resolved layoutDirection is LTR, and ALIGN_RIGHT otherwise.


The current Litho implementation does not support the last 2: **viewEnd** and **viewStart**.

astreet commented:
viewStart and viewEnd (as names at least) won't mean much in Litho because we don't have guarantees around when you will have a View and when you won't: what's the actual contract here? For example, it doesn't look like we use any View data in your implementation.

Otherwise, we'll need to make this non-breaking for now -- at this point I'm not sure if that means just add some ints to the existing list of possible alignments (what I'm favoring) or creating a new api and deprecating the old one (which is a lot more work).

vinc3m1 commented:
`viewStart/End` could be renamed `layoutStart/End` for better alignment with Litho, basically the text aligns with the direction of the layout it's inside.

The problem with the existing API is that it uses [Layout.Alignment](https://developer.android.com/reference/android/text/Layout.Alignment) which is an enum and not extensible, so it'll be hard to extend without breaking changes.

ironmanTony commented:
in Text the right to left is not correct.

![image](https://user-images.githubusercontent.com/5860185/63083387-3013d700-bf7b-11e9-8622-5cf1cbd673d7.png)


```
        ComponentContext c = new ComponentContext(this);
        View view = LithoView.create(c,
                Column.create(c).heightDip(40)
                        .child(Text.create(c).flex(1).text(" forever faster سرعة الي الابد").textSizeSp(18)
                                .layoutDirection(YogaDirection.RTL)
                                .backgroundColor(Color.GREEN)
                                .build())
                        .child(Text.create(c).flex(1).text("الحقائب المحبوكة للصيف").textSizeSp(18).build())
                        .build());
        setContentView(view);
```

vinc3m1 commented:
@ironmanTony Android's default text alignment behavior depends on the content of the text, specifically the first strongly directional character, in this case since the string starts with English, the system aligns to the left. This behavior is unrelated to the PR at hand though, so if you have follow up questions/discussion feel free to open up another issue to discuss.

vinc3m1 commented:
@astreet author has already responded before you tagged 😄, would `layoutStart` & `layoutEnd` be sufficient changes to make this acceptable? It doesn't really change the fact that it's a breaking change though. The current API uses the [Layout.Alignment](https://developer.android.com/reference/android/text/Layout.Alignment) enum that we can't really extend, and also takes the API name we want...

Ideally the existing API would be named `layoutAlignment` and the new one could be `textAlignment`. If we can come up with a better name for the new API, we could deprecate the existing one and map the values over. Maybe something like `realTextAlignment` 😛

astreet commented:
Whoops, sorry about that :) Yes, layoutStart/End sounds fine, just wanted to make sure we weren't tying this to Views.

For the breaking aspect, can we potentially just add new ints as opposed to moving to use the enums?

vinc3m1 commented:
I've made this a non-breaking change with a less-than-ideal prop naming scheme, but should ease migrations.

vinc3m1 commented:
Only failing test is due to: https://github.com/facebook/litho/commit/11d8bffda27f4647bd14a9fdba7f187e7f41ea68#r34873793

astreet commented:
Re: enum, I see. Can we add this as a new prop as you've done and just mark the old one as deprecated? The new prop would need to have a default that lets us know it's not set so we can fall back to the old prop.

vinc3m1 commented:
@astreet yeah I've updated the code to do exactly that already, currently using a nullable `Integer` to know if it's set or not. I've also switched to IntDef over Enum, but would maybe prefer an Enum for a stricter API? Not a strong preference so let me know.

astreet commented:
Ah, I didn't notice -- changes look good to me. Can you add a test to TextSpecTest and I'll go ahead and re-import?

astreet commented:
I'm running our screenshot tests on it now -- code looks good but would like a test for it (e.g. in TextSpecTest)

vinc3m1 commented:
yeah, I just rebased, didn't get a chance to change to enums and add tests
yet

On Mon, Sep 9, 2019 at 7:28 AM Andy Street <notifications@github.com> wrote:

> I'm running our screenshot tests on it now -- code looks good but would
> like a test for it (e.g. in TextSpecTest)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/facebook/litho/pull/535?email_source=notifications&email_token=AADZZJARPOQRLBDQIB2HRPLQIZMPBA5CNFSM4HFXHEO2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6HY32Y#issuecomment-529501675>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AADZZJHDCQONHPSL4MDP433QIZMPBANCNFSM4HFXHEOQ>
> .
>

vinc3m1 commented:
tests still TBD, will get them tomorrow

vinc3m1 commented:
Alright unit tests done, now we're finally ready for review!

vinc3m1 commented:
any udpates @astreet? I assume you're fixing internal lints/tests? I have some people bugging me about when this gets merged so just an estimate would be helpful, thanks!

astreet commented:
Hopefully early next week? I'm trying to get screenshot tests to actually work on the diff to make sure it won't break anything internally

colriot commented on litho-it/src/test/java/com/facebook/litho/widget/TextSpecTest.java:
```diff
@@ -38,11 +38,15 @@
 import org.junit.runner.RunWith;
 import org.robolectric.RuntimeEnvironment;
 
+import javax.annotation.Nullable;
```
Can you, please, use AndroidX annotation here, so that we won't need to import LITHO_JSR_TARGET as a dep in BUCK?

vinc3m1 replied:
lol you're gonna have to ask @astreet for these fixes now :P

colriot replied:
Yeah, it's ok. We pulled this in order to test in asap :)

colriot commented on litho-it/src/test/java/com/facebook/litho/widget/TextSpecTest.java:
```diff
@@ -345,4 +350,249 @@ private static Layout setupWidthTestTextLayout() {
+        YogaDirection.LTR,
+        null,
+        TextAlignment.RIGHT))
+        .isEqualTo(Layout.Alignment.ALIGN_NORMAL);
```
Javadoc for `LEFT` and `RIGHT` says that they don't respect text/locale prefs. Here in the test `RIGHT` doesn't respect `YogaDiraction`, but does respect Arabic's RTL nature. 
Am I wrong or this doesn't correspond to the javadoc?

vinc3m1 replied:
RIGHT means always right, but since the only public alignment enums we can use are NORMAL and OPPOSITE, we have to choose based on the directionality of the text itself. 

e.g. for RIGHT:
NORMAL for RTL text
OPPOSITE for LTR text

to make sure that the alignment is always RIGHT, regardless of text direction.

astreet replied:
I think Sergey's referring to the docs for TextAlignment that was added that say "Align the text to the left, ignoring the text or locale preferences for LTR/RTL.", but that doesn't seem to be reflected in these tests

vinc3m1 replied:
Nah I understood him perfectly, and I’m trying to explain that this test verified exactly what the docs intend.

English text + ALIGN_OPPOSITE == right aligned
Arabic text + ALIGN_NORMAL == right aligned

^ that’s what this test checks for

vinc3m1 replied:
Yoga direction is completely ignored, as it should be.

astreet replied:
Ah gotcha, it's like... double-factoring it in and therefore canceling it out :P

vinc3m1 replied:
Yeah it’s pretty confusing lol. Sucks that ALIGN_RIGHT and ALIGN_LEFT are hidden in the enum... otherwise we could just use those directly.

colriot replied:
Thanks for clarification, Vince. This double-factoring was confusing, was expecting that NORMAL and OPPOSITE don't change depending on the text locale.
